File: apps/analyzer/llm/token_utils.py
from typing import List

try:
    from transformers import AutoTokenizer
    transformers_available = True
except ImportError:
    transformers_available = False

_tokenizer_cache = {}

# Mapping des modèles connus → tokenizer HuggingFace
model_to_tokenizer = {
    "Llama-3.1-8B-Instruct": "meta-llama/Meta-Llama-3-8B-Instruct",
    "Meta-Llama-3_3-70B-Instruct": "meta-llama/Meta-Llama-3-70B-Instruct",
    "gpt-3.5-turbo": "gpt2",
    "gpt2": "gpt2"
}

def estimate_token_count(messages: List[dict], model_name: str = "") -> int:
    """
    Estimate token count for a list of prompt messages.
    Uses HuggingFace tokenizer if available and compatible, else fallback to naive estimate.
    """
    if not transformers_available:
        print("[Token Estimation] transformers not installed, using fallback.")
        return naive_token_estimate(messages)

    tokenizer_name = model_to_tokenizer.get(model_name)
    if not tokenizer_name:
        print(f"[Token Estimation] No tokenizer mapping for model '{model_name}', using fallback.")
        return naive_token_estimate(messages)

    try:
        if tokenizer_name not in _tokenizer_cache:
            _tokenizer_cache[tokenizer_name] = AutoTokenizer.from_pretrained(tokenizer_name)
        tokenizer = _tokenizer_cache[tokenizer_name]
    except Exception as e:
        print(f"[Token Estimation] Failed to load tokenizer for {model_name}: {e}")
        return naive_token_estimate(messages)

    total_tokens = 0
    for msg in messages:
        content = msg.get("content", "")
        role = msg.get("role", "")
        tokens = tokenizer.encode(role + ": " + content, add_special_tokens=False)
        total_tokens += len(tokens)
    return total_tokens

def naive_token_estimate(messages: List[dict]) -> int:
    """
    Fallback if tokenizer is unavailable. Rough estimation: 1 token ≈ 4 chars
    """
    total_chars = sum(len(msg.get("role", "")) + len(msg.get("content", "")) for msg in messages)
    return total_chars // 4

File: apps/backend/src/analysis/llm/token-utils.ts
import type { PreTrainedTokenizer } from '@xenova/transformers';

const modelToTokenizer: Record<string, string> = {
	'Llama-3.1-8B-Instruct': 'meta-llama/Meta-Llama-3-8B-Instruct',
	'Meta-Llama-3_3-70B-Instruct': 'meta-llama/Meta-Llama-3-70B-Instruct',
	'gpt-3.5-turbo': 'gpt2', // approximation
	'gpt2': 'gpt2'
};

// Modèles qui nécessitent l'estimation naïve (pas de tokenizer)
const modelsUsingNaiveEstimation = [
	'Meta-Llama-3_3-70B-Instruct'
];

const tokenizerCache: Record<string, PreTrainedTokenizer> = {};

type Message = {
  role: string;
  content: string;
};

async function loadTokenizer(modelName: string): Promise<PreTrainedTokenizer | null> {
	// Si le modèle nécessite l'estimation naïve, on retourne null directement
	if (modelsUsingNaiveEstimation.includes(modelName)) {
		console.warn(`[TokenUtils] Using naive estimation for "${modelName}" (private model)`);
		return null;
	}

	const identifier = modelToTokenizer[modelName];
	if (!identifier) return null;

	if (tokenizerCache[identifier]) return tokenizerCache[identifier];

	try {
		const { AutoTokenizer } = await import('@xenova/transformers');
		const tokenizer = await AutoTokenizer.from_pretrained(identifier);
		tokenizerCache[identifier] = tokenizer;
		return tokenizer;
	} catch (err) {
		console.warn(`[TokenUtils] Failed to load tokenizer for model "${modelName}"`, err);
		console.warn(`[TokenUtils] Using fallback estimation for "${modelName}"`);
		return null;
	}
}

function naiveTokenEstimate(messages: Message[]): number {
	const charCount = messages.reduce(
		(sum, msg) => sum + msg.role.length + msg.content.length,
		0
	);
	return Math.floor(charCount / 4);
}

export async function estimateTokenCount(
	messages: Message[],
	modelName: string
	): Promise<number> {
	const tokenizer = await loadTokenizer(modelName);
	if (!tokenizer) return naiveTokenEstimate(messages);

	let total = 0;
	for (const msg of messages) {
		const encoded = await tokenizer.encode(`${msg.role}: ${msg.content}`);
		total += encoded.length;
	}
	return total;
}
